### DISCUSSION TITLE
capababilties for dialog plugin and remote urls

### DISCUSSION THREAD

**Discussion by mcaox** [tauri-apps/tauri#11622] [Q&A] [answered]
# Use case

We use tauri as a launcher of some webapps with the following sequence
- main 
  - set webapp arguments depending on command line arguments
  - searchs for an available port -> PORT
  - spawns a local server serving the webapp on that PORT
  - eval a javascript statement `window.location.replace` in the main window 
    - for instance `window.location.replace('http://localhost:53797')`, when PORT is 53797
    - it is performed through the rust statement `let _ = main_window.eval(...);`
  - the redirection works
  - the web app use the following statement through the javascript api
    - `window.__TAURI__.dialog['open']({defaultPath:'something',directory: true})`

# tauri v1
## issue
  - no issue, the launcher works as expected and dialog is functional
 
# tauri v2
## issue
  - the following error is thrown
    - ```dialog.open not allowed on window "main", webview "main", URL: http://localhost:53797/allowed on: [windows: "main", webviews: "main", URL: http://127.0.0.1/*], [windows: "main", webviews: "main", URL: http://localhost/*], [windows: "main", URL: local], [windows: "main", URL: local]referenced by: capability: main-capability, permission: allow-open || capability: main-capability, permission: allow-open || capability: migrated, permission: allow-open || capability: migrated, permission: allow-open``` 

### tauri.conf.json
```
{
  "$schema": "../node_modules/@tauri-apps/cli/schema.json",
  "build": {
    "beforeBuildCommand": "",
    "beforeDevCommand": "",
    "frontendDist": "../ui"
  },
  "bundle": {
    "active": true,
    "targets": "all",
    "icon": [
      "icons/32x32.png",
      "icons/128x128.png",
      "icons/128x128@2x.png",
      "icons/icon.icns",
      "icons/icon.ico"
    ],
    "shortDescription": "",
    "longDescription": ""
  },
  "productName": "serverlauncher",
  "identifier": "com.demoapp.serverlauncher",
  "app": {
    "withGlobalTauri": true,
    "windows": [
      {
        "fullscreen": false,
        "height": 600,
        "resizable": true,
        "title": "serverlauncher",
        "width": 800,
        "label": "main"
      }
    ],
    "security": {}
  }
}
```
### capabilities/default.json
```
{
    "$schema": "./schemas/desktop-schema.json",
    "identifier": "main-capability",
    "description": "Capability for the main window",
    "windows": ["main"],
    "webviews": ["main"],
    "local":false,
    "remote":{"urls": ["http://127.0.0.1","http://localhost","http{s}?://localhost(:\d+)?"]},
    "permissions": [
      "core:path:default",
      "core:event:default",
      "core:window:default",
      "core:app:default",
      "core:resources:default",
      "core:menu:default",
      "core:tray:default",
      "core:window:allow-set-title",
      "dialog:allow-ask",
      "dialog:allow-confirm",
      "dialog:allow-message",
      "dialog:allow-open",
      "dialog:allow-save"
    ]
  }
```

# question

We tried different combinations for **remote** and **local** and followed [https://infra.spec.whatwg.org](vscode-file://vscode-app/c:/Users/maxime.conti/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html)

We both list and didn't list the capability path in the tauri.conf.json file.

The issue probably results from a mistake in the migration from tauri v1 to tauri v2. We didn't manage to locate this mistake. Any idea ?

**Comment by FabianLars** [comment]
Did you also try (temporarily) hardcoding the port already? From what you wrote, including the posted files, an issue with the url validation in the capability check is what seems most likely to me 🤔

**Comment by mcaox** [comment, marked as answer]
Hi Fabian,

Thank you for your tip.

Indeed when I hardcode the port in the capabilty definition it works. It means the issue is on the url syntax ...

``` json
    "remote":{"urls": ["http{s}?://localhost:8080"]},
```

I read again [patterns specification](https://urlpattern.spec.whatwg.org/#pattern-string) and tried several combinations assuming port was a specific "part". I found any functional.

Finally I tried this really simple syntax, that works whereas it doesn't really follow the [patterns specification](https://urlpattern.spec.whatwg.org/#pattern-string)  (from my understanding) :

``` json
    "remote":{"urls": ["http{s}?://localhost:*/*"]},
```
